src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentitySource.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Utils;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Core;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal abstract class ManagedIdentitySource
+    {
+        internal const string AuthenticationResponseInvalidFormatError = "Invalid response, the authentication response was not in the expected format.";
+        internal const string UnexpectedResponse = "Managed Identity response was not in the expected format. See the inner exception for details.";
+        internal RequestContext _requestContext;
+
+        protected ManagedIdentitySource(RequestContext requestContext)
+        {
+            _requestContext = requestContext;
+        }
+
+        protected internal string ClientId { get; }
+
+        public virtual async Task<ManagedIdentityResponse> AuthenticateAsync(AppTokenProviderParameters parameters, CancellationToken cancellationToken)
+        {
+            ManagedIdentityRequest request = CreateRequest(parameters.Scopes.ToArray());
+            
+            var response =
+            HttpMethod.Get.Equals(request.Method) ?
+            await _requestContext.ServiceBundle.HttpManager.SendGetAsync(request.UriBuilder.Uri, request.Headers, _requestContext.Logger, useManagedIdentity: true, cancellationToken: cancellationToken).ConfigureAwait(false) :
+            await _requestContext.ServiceBundle.HttpManager.SendPostAsync(request.UriBuilder.Uri, request.Headers, request.BodyParameters, _requestContext.Logger, useManagedIdentity: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return HandleResponse(parameters, response);
+            
+        }
+
+        protected virtual ManagedIdentityResponse HandleResponse(
+            AppTokenProviderParameters parameters,
+            HttpResponse response)
+        {
+            try
+            {
+                if (response.StatusCode == System.Net.HttpStatusCode.OK)
+                {
+                    return GetSuccessfulResponse(response);
+                }
+
+                throw MsalServiceExceptionFactory.FromManagedIdentityResponse(MsalError.ManagedIdentityRequestFailed, response);
```

Comment:
Pls make sure to add more logging statements everywhere, especially error conditions.